### PR TITLE
Unit10 hometask

### PR DESCRIPTION
**Pull request by letztersommer**
Perform and verify sorting on the Inventory page:
- sort name AZ
- sort name ZA
- sort price HL
- sort price LH

### CODE REVIEW

File: src/tests/inventorySorting.spec.js [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+        await app.login.performLogin('standard_user', 'secret_sauce');
+
+        expect(await app.inventory.inventoryItems.count()).toBeGreaterThanOrEqual(1);
+
+        const productNamesBeforeSort = await app.inventory.inventoryItems.allTextContents();
+
+        await app.inventory.sortProductNameAZ();
+        const productNamesAfterSortAZ = await app.inventory.inventoryItems.allTextContents();
+        const sortedNamesAZ = [...productNamesBeforeSort].sort();
+        expect(sortedNamesAZ).toEqual(productNamesAfterSortAZ);
```
Comment (kirstaVova) [inline src/tests/inventorySorting.spec.js:18]: You can also enable an additional check to make sure the array is not empty before sorting

File: src/tests/inventorySorting.spec.js [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+        expect(await app.inventory.inventoryItems.count()).toBeGreaterThanOrEqual(1);
+
+        const productPriceBeforeSort = await app.inventory.inventoryItemsPrice.allTextContents();
+        
+        await app.inventory.sortProductPriceLoHi();
+        const productPriceAfterSortLoHi = await app.inventory.inventoryItemsPrice.allTextContents();
+        const cleanedProductPriceAfterSortLoHi = productPriceAfterSortLoHi.map(price => parseFloat(price.replace(/\$/g, '')));
+        const sortedPricesLoHi = [...productPriceBeforeSort].map(price => parseFloat(price.replace(/\$/g, ''))).sort((a, b) => a - b);
+        expect(sortedPricesLoHi).toEqual(cleanedProductPriceAfterSortLoHi);
```
Comment (ihorBohdvn) [inline src/tests/inventorySorting.spec.js]: Good work!)

**Comment by ihorBohdvn** [comment]
Very nice job, you're doing great! I have just a couple more comments:
1)Put the common logic in beforeEach: Login and navigation are repeated in every test.
2)Clearing and sorting prices is duplicated, so put it in a function.

**Comment by letztersommer** [comment]
_2)Clearing and sorting prices is duplicated, so put it in a function._

I did not understand how to make 2) task completely. I made a function for cleaning prices, but couldn`t make the same for sorting, couse it is not the same sort for both tests. So, I would be grateful for a hint here.

**Comment by ihorBohdvn** [comment]
> _2)Clearing and sorting prices is duplicated, so put it in a function._
> 
> I did not understand how to make 2) task completely. I made a function for cleaning prices, but couldn`t make the same for sorting, couse it is not the same sort for both tests. So, I would be grateful for a hint here.

1) The logic for clearing prices (removing $ and converting to a number) should be placed in a separate function.
2) Sorting can also be optimized by creating a single function that accepts an order parameter (asc or desc).
3) In test methods, it is better to call the price clearing function directly from objects.

import { expect } from '@playwright/test';
import { test } from '../fixtures/base';

// function for cleaning prices
const cleanPrices = (prices) => prices.map(price => parseFloat(price.replace(/\$/g, '')));

//  function for sorting prices
const sortPrices = (prices, order) => {
    const cleanedPrices = cleanPrices(prices);
    return order === 'asc' ? cleanedPrices.sort((a, b) => a - b) : cleanedPrices.sort((a, b) => b - a);
};

test.beforeEach(async (
    /** @type {{ app: import('../pages/Application').Application }} */{ app },
) => {
    await app.login.navigate();
    await app.login.performLogin('standard_user', 'secret_sauce');

    expect(await app.inventory.inventoryItems.count()).toBeGreaterThanOrEqual(1);
});

test.describe('Perform and verify sorting on the Inventory page', () => {
    test('Sorting products name A-Z', async (
        /** @type {{ app: import('../pages/Application').Application }} */{ app },
    ) => {
        const productNamesBeforeSort = await app.inventory.inventoryItems.allTextContents();

        await app.inventory.sortProductNameAZ();
        const productNamesAfterSortAZ = await app.inventory.inventoryItems.allTextContents();
        const sortedNamesAZ = [...productNamesBeforeSort].sort();
        expect(productNamesAfterSortAZ).toEqual(sortedNamesAZ);
    });

    test('Sorting products name Z-A', async (
        /** @type {{ app: import('../pages/Application').Application }} */{ app },
    ) => {
        const productNamesBeforeSort = await app.inventory.inventoryItems.allTextContents();

        await app.inventory.sortProductNameZA();
        const productNamesAfterSortZA = await app.inventory.inventoryItems.allTextContents();
        const sortedNamesZA = [...productNamesBeforeSort].sort().reverse();
        expect(productNamesAfterSortZA).toEqual(sortedNamesZA);
    });

    test('Sorting products prices Low - High', async (
        /** @type {{ app: import('../pages/Application').Application }} */{ app },
    ) => {
        const productPriceBeforeSort = await app.inventory.inventoryItemsPrice.allTextContents();
        const cleanedProductPriceBefore = cleanPrices(productPriceBeforeSort);

        await app.inventory.sortProductPriceLoHi();

        const productPriceAfterSortLoHi = await app.inventory.inventoryItemsPrice.allTextContents();
        const cleanedProductPriceAfter = cleanPrices(productPriceAfterSortLoHi);

        const sortedPrices = sortPrices(cleanedProductPriceBefore, 'asc');
        expect(cleanedProductPriceAfter).toEqual(sortedPrices);
    });

    test('Sorting products prices High - Low', async (
        /** @type {{ app: import('../pages/Application').Application }} */{ app },
    ) => {
        const productPriceBeforeSort = await app.inventory.inventoryItemsPrice.allTextContents();
        const cleanedProductPriceBefore = cleanPrices(productPriceBeforeSort);

        await app.inventory.sortProductPriceHiLo();

        const productPriceAfterSortHiLo = await app.inventory.inventoryItemsPrice.allTextContents();
        const cleanedProductPriceAfter = cleanPrices(productPriceAfterSortHiLo);

        const sortedPrices = sortPrices(cleanedProductPriceBefore, 'desc');
        expect(cleanedProductPriceAfter).toEqual(sortedPrices);
    });
});

**Review by ihorBohdvn — APPROVED** [review]

File: src/pages/Inventory.page.js [hunk 3]
```diff
@@ -9,7 +9,27 @@ export class InventoryPage extends BaseSwagLabPage {
...
+        await this.sortProductButton.selectOption('za');
+    }
+
+    async sortProductPriceLoHi() {
+        await this.sortProductButton.selectOption('lohi');
+    }
+
+    async sortProductPriceHiLo() {
+        await this.sortProductButton.selectOption('hilo');
+    }
```
Comment (batmanastasiya) [inline src/pages/Inventory.page.js:34]: all these functions can be replaced by one: 
```
async sortProductBy(option) {
await this.sortProductButton.selectOption(option);
}
```

File: src/tests/inventorySorting.spec.js [hunk 4]
```diff
@@ -0,0 +1,69 @@
+import { expect } from '@playwright/test';
+import { test } from '../fixtures/base';
+
+const cleanPrices = (prices) => prices.map(price => parseFloat(String(price).replace(/$/g, '')));
+
+const sortPrices = (prices, order) => {
+    const cleanedPrices = cleanPrices(prices);
+    return order === 'asc' ? cleanedPrices.sort((a, b) => a - b) : cleanedPrices.sort((a, b) => b - a);
+};
+
```
Comment (batmanastasiya) [inline src/tests/inventorySorting.spec.js:10]: It is a great idea to create those. Usually, we are using a separate file to store the helper functions like that

File: src/tests/inventorySorting.spec.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+test.beforeEach(async (
+    /** @type {{ app: import('../pages/Application').Application }} */{ app },
+) => {
+    await app.login.navigate();
+    await app.login.performLogin('standard_user', 'secret_sauce');
```
Comment (batmanastasiya) [inline src/tests/inventorySorting.spec.js:15]: Please, create a separate file with credentials and use them inside all tests you need. 
To use it like: 
```
await app.login.performLogin(user.username, user.password);
```

File: src/tests/inventorySorting.spec.js [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+test.describe('Perform and verify sorting on the Inventory page', () => {
+    test('Sorting products name A-Z', async (
+        /** @type {{ app: import('../pages/Application').Application }} */{ app },
+    ) => {
+        const productNamesBeforeSort = await app.inventory.inventoryItems.allTextContents();
+
+        await app.inventory.sortProductNameAZ();
+        const productNamesAfterSortAZ = await app.inventory.inventoryItems.allTextContents();
+        const sortedNamesAZ = [...productNamesBeforeSort].sort();
```
Comment (batmanastasiya) [inline src/tests/inventorySorting.spec.js:28]: Also can be extracted to helper functions

**Comment by batmanastasiya** [comment]
Sorry, I missed the fact that the PR was already approved

**Pull request merged**